src/main/java/sandbox/JdbcSqlStorageIterator.java

```diff
@@ -1,28 +1,147 @@
+
 package sandbox;
 
+import org.commcare.modern.database.DatabaseHelper;
+import org.commcare.modern.database.TableBuilder;
+import org.javarosa.core.services.storage.IStorageIterator;
 import org.javarosa.core.services.storage.Persistable;
+import org.javarosa.core.services.storage.StorageModifiedException;
+import org.javarosa.core.util.ArrayUtilities;
 
-import java.util.ArrayList;
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.*;
 
 /**
- * Simple implemenation of IStorageIterator that is actually just backed by an ArrayList
- * since we don't want to deal with Connections or ResultSets lingering around.
+ * Iterator that only processes id indexes, and not real data.
  *
+ * @author ctsims
  * @author wspride
  */
-public class JdbcSqlStorageIterator <E extends Persistable> extends AbstractSqlIterator {
+public class JdbcSqlStorageIterator<T extends Persistable> implements IStorageIterator<T>, Iterator<T> {
+
+    protected PreparedStatement preparedStatement;
+    protected ResultSet resultSet;
+    private final Set<String> metaDataIndexSet;
+    SqliteIndexedStorageUtility<T> storage;
+    private HashMap<String, Integer> metaDataColumnMap = new HashMap<>();
+
+    public JdbcSqlStorageIterator(PreparedStatement preparedStatement,
+                                  ResultSet resultSet,
+                                  SqliteIndexedStorageUtility<T> storage,
+                                  String[] metaDataIndexSet) {
+        this.resultSet = resultSet;
+        this.storage = storage;
+        this.metaDataIndexSet = new HashSet<>(ArrayUtilities.toVector(metaDataIndexSet));
+        this.preparedStatement = preparedStatement;
+        try {
+            resultSet.next();
```

Comment:
If this returns false() the resultSet is empty, but it will never get closed, because hasMore() doesn't close a resultSet that is after the last element. 

Either this should close the resultSet if its empty, or hasMore() should close resultSets that have iterated past their last element.,